src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOptimizedNullableOperators.cs

```diff
@@ -2414,5 +2414,562 @@ .maxstack  2
+                  IL_002a:  ret
+                }
+                """);
+            verifier = CompileAndVerify(source, options: TestOptions.DebugExe, expectedOutput: expectedOutput);
+            verifier.VerifyDiagnostics();
+            verifier.VerifyMethodBody("C.M", """
+                {
+                  // Code size       47 (0x2f)
+                  .maxstack  3
+                  // sequence point: {
+                  IL_0000:  nop
+                  // sequence point: Write(b ?? true);
+                  IL_0001:  ldarga.s   V_0
+                  IL_0003:  call       "bool bool?.GetValueOrDefault()"
+                  IL_0008:  ldarga.s   V_0
+                  IL_000a:  call       "bool bool?.HasValue.get"
+                  IL_000f:  ldc.i4.0
+                  IL_0010:  ceq
+                  IL_0012:  or
+                  IL_0013:  call       "void C.Write(bool)"
+                  IL_0018:  nop
+                  // sequence point: Write(b ?? false);
+                  IL_0019:  ldarga.s   V_0
+                  IL_001b:  call       "bool bool?.GetValueOrDefault()"
+                  IL_0020:  call       "void C.Write(bool)"
+                  IL_0025:  nop
+                  // sequence point: System.Console.Write(' ');
+                  IL_0026:  ldc.i4.s   32
+                  IL_0028:  call       "void System.Console.Write(char)"
+                  IL_002d:  nop
+                  // sequence point: }
+                  IL_002e:  ret
+                }
+                """);
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/66431")]
+        public void NullableConstant_Boolean()
+        {
+            var source = """
+                C.M(true);
+                C.M(false);
+                C.M(null);
+
+                class C
+                {
+                    static void Write(bool b) => System.Console.Write(b ? 1 : 0);
+
+                    public static void M(bool? b)
+                    {
+                        Write(b == true);
+                        Write(true == b);
+                        Write(b == false);
+                        Write(false == b);
+                        System.Console.Write('-');
+                        Write(b != true);
+                        Write(true != b);
+                        Write(b != false);
+                        Write(false != b);
+                        System.Console.Write('-');
+                        Write(!(b == true));
+                        Write(!(b != true));
+                        Write(!!(b == true));
+                        System.Console.Write(' ');
+                    }
+                }
+                """;
+            var expectedOutput = "1100-0011-011 0011-1100-100 0000-1111-100";
+            var verifier = CompileAndVerify(source, options: TestOptions.ReleaseExe, expectedOutput: expectedOutput);
+            verifier.VerifyDiagnostics();
+            verifier.VerifyMethodBody("C.M", """
```

Comment:
>verifier.VerifyMethodBody("C.M", """ [](http://example.com/codeflow?start=12&length=36)

I understand that you are trying to rely on sequence points to match portions of IL to source code. However, this approach is fragile because any change at the beginning of the method will cause changes all the way through the rest of the IL and it will be very hard to see what changes are "interesting" and which are not. Reviewing IL for long methods (even with sequence points) is a major pain. So, please split the big method into several focused methods and assert IL for each method separately (dropping the sequence points, if it is not your goal to test them). Separating scenarios into individual unit-tests will be even better, this way a failure will not affect tests for scenarios that aren't broken.
This comment applies to other added tests too.